### PR TITLE
samples: mesh: nrf52: removed bugs in Gen. level Move GET & Publish

### PR DESCRIPTION
**Pull request by vikrant8052**
Due to recent changes some bugs had introduced in Gen. Level Move
GET & Publish functionalities. Now separatly introduced
gen_level_move_get() and gen_level_move_publisher() to resolved it.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

### CODE REVIEW

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 1]
```diff
@@ -679,6 +648,63 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+		/* This is as per PTS requirement */
+		net_buf_simple_add_u8(msg, 0x3F);
+	}
+
+	if (bt_mesh_model_send(model, ctx, msg, NULL, NULL)) {
+		printk("Unable to send GEN_LEVEL_SRV Status response\n");
+	}
+}
+
+static void gen_level_move_publisher(struct bt_mesh_model *model)
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: I know this is not the only function with the issue, but function names should have a verb as the last component. Here you have a noun. I.e. the name should be gen_level_move_publish()

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 2]
```diff
@@ -679,6 +648,63 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+{
+	struct net_buf_simple *msg = model->pub->msg;
+	struct generic_level_state *state = model->user_data;
+
+	if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED) {
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]:
```
if (model->pub->addr == BT_MESH_ADDR_UNASSIGNED) {
         ...
}
```

Helps you eliminate unnecessary indentation, which in turn eliminates unnecessary breaks of long lines (I know this is not an issue for this particular function, but it's good to get into the habit and be consistent with this kind of function flow).

**Review by jhedberg — CHANGES_REQUESTED** [review]
A few minor things

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 2 above)
Comment (jhedberg) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: (edited above comment - "..." was supposed to be "return")

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 1 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 2 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 3]
```diff
@@ -679,6 +650,64 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+	if (state->transition->counter) {
+
+		if (state->last_delta < 0) {
+			net_buf_simple_add_le16(msg, INT16_MIN);
+		} else if (state->last_delta > 0) {
+			net_buf_simple_add_le16(msg, INT16_MAX);
+		}
+
+		/* This is as per PTS requirement */
+		net_buf_simple_add_u8(msg, 0x3F);
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: You might want to have a define for 0x3F instead of having it as a magic constant here. As per section 3.1.3.1 in the Mesh Model spec it means "The value is unknown". Saying that it's a PTS requirement isn't helpful since the PTS only follows the Test Specification which in turn follows the Model Specification.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 4]
```diff
@@ -679,6 +650,64 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+	bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_GEN_LEVEL_STATUS);
+	net_buf_simple_add_le16(msg, state->level);
+
+	if (state->transition->counter) {
+
+		if (state->last_delta < 0) {
+			net_buf_simple_add_le16(msg, INT16_MIN);
+		} else if (state->last_delta > 0) {
+			net_buf_simple_add_le16(msg, INT16_MAX);
+		}
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: What if last_delta is 0? I that case you'd create a corrupted message without a Target Level field.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 5]
```diff
@@ -679,6 +650,64 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+			       struct net_buf_simple *buf)
+{
+	struct net_buf_simple *msg = NET_BUF_SIMPLE(2 + 5 + 4);
+	struct generic_level_state *state = model->user_data;
+
+	bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_GEN_LEVEL_STATUS);
+	net_buf_simple_add_le16(msg, state->level);
+
+	if (state->transition->counter) {
+
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: minor: redundant empty line

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 6]
```diff
@@ -679,6 +650,64 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+	if (model->pub->addr == BT_MESH_ADDR_UNASSIGNED) {
+		return;
+	}
+
+	bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_GEN_LEVEL_STATUS);
+	net_buf_simple_add_le16(msg, state->level);
+
+	if (state->transition->counter) {
+
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: minor: redundant empty line

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 7]
```diff
@@ -679,6 +650,64 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+		if (state->last_delta < 0) {
+			net_buf_simple_add_le16(msg, INT16_MIN);
+		} else if (state->last_delta > 0) {
+			net_buf_simple_add_le16(msg, INT16_MAX);
+		}
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: See my other related comment

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 8]
```diff
@@ -679,6 +650,64 @@ static void gen_delta_set(struct bt_mesh_model *model,
...
+
+		/* This is as per PTS requirement */
+		net_buf_simple_add_u8(msg, 0x3F);
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: See my other related comment.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 9]
```diff
@@ -961,21 +991,22 @@ static void gen_onpowerup_status(struct bt_mesh_model *model,
...
 
-	if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED) {
-		int err;
+	if (model->pub->addr == BT_MESH_ADDR_UNASSIGNED) {
+		return;
+	}
 
-		bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_2(0x82, 0x12));
-		net_buf_simple_add_u8(msg, state->onpowerup);
+	bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_2(0x82, 0x12));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Why do you have these magic 0x82, 0x12 values here when in other places you have proper defines for the opcodes (e.g. `BT_MESH_MODEL_OP_GEN_LEVEL_STATUS`). Please use proper descriptive defines instead of magic values. Otherwise you're just making it hard to follow the code and understand exactly which message is being sent.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 4 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: when client sends delta = 0 means stop the transition.
Plus it will also set state->transition->counter = 0

That means following lines would not execute in that conditions

```
if (state->transition->counter) {
		if (state->last_delta < 0) {
			net_buf_simple_add_le16(msg, INT16_MIN);
		} else if (state->last_delta > 0) {
			net_buf_simple_add_le16(msg, INT16_MAX);
		}

		/* This is as per PTS requirement */
		net_buf_simple_add_u8(msg, UNKNOWN_VALUE);
}

```

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 7 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: As mentioned earlier, these lines would not get executed when
delta = 0 because when it is zero  `state->transition->counter` become zero.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 9 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: I will take care of this in separate PR.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 3 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 6 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 5 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 8 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 9 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Done !!

**Comment by jhedberg** [comment]
recheck

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 7 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 4 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Done !!

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: done !!

**Comment by vikrant8051** [comment]
All suggested changes has done.

**Review by jhedberg — APPROVED** [review]
